### PR TITLE
vec proto and useful macros

### PR DESCRIPTION
**Pull request by SchrodingerZhu**

### CODE REVIEW

File: cc_assert.h [hunk 1]
```diff
@@ -35,4 +35,29 @@ extern void cclib_assert_impl(int value,
 } /* extern "C" */
 #endif
 
+
+#ifndef NDEBUG
+#   define CC_ASSUME(x) CC_ASSERT(x)
```
Comment (ICEYSELF) [inline cc_assert.h:40]: What's the basic difference between assert and assume?

File: cc_defs.h [hunk 2]
```diff
@@ -15,8 +15,44 @@ typedef CCLIB_SIZE_TYPE CCTY(cc_size);
...
+#   define CC_PURE             // do nothing
+#   define CC_UNREACHABLE_HINT // do nothing
+#else
+#   define CC_FAST_PATH        inline
+#   define CC_SLOW_PATH        // do nothing
+#   define CC_UNLIKELY(x)      !!(x)
+#   define CC_LIKELY(x)        !!(x)
+#   define CC_PURE             // do nothing 
+#   define CC_UNREACHABLE_HINT 
+#endif
```
Comment (ICEYSELF) [inline cc_defs.h:54]: Perhaps we may provide an cc_defs.user.h for user to define more stuffs.

**Review by ICEYSELF — COMMENTED** [review]
Approve.

File: cc_assert.h (hunk 1 above)
Comment (SchrodingerZhu) [reply cc_assert.h:40]: assume also gives optimizer a hint.

File: cc_defs.h (hunk 2 above)
Comment (SchrodingerZhu) [reply cc_defs.h:54]: maybe. but it is just compiler specific？

Comment (ICEYSELF) [reply cc_defs.h:54]: But there are compilers like sdcc, armcc, keilc, tcc, ... and so on.

**Review by ICEYSELF — APPROVED** [review]

File: cc_defs.h (hunk 2 above)
Comment (SchrodingerZhu) [reply cc_defs.h:54]: will change it. How about:
```c++
#ifndef XXX
#define XXX YYY
#endif
```
and in `.user.h`:
```c++
#define XXX YYY
// ....
```

Comment (ICEYSELF) [reply cc_defs.h:54]: Yes and this is what I'd like to do

**Comment by ICEYSELF** [comment]
Is this pull request finished?

**Comment by SchrodingerZhu** [comment]
Yes.

**Pull request merged**